src/libraries/System.Security.Cryptography/tests/X509Certificates/X509FilesystemTests.Windows.cs

```diff
@@ -0,0 +1,464 @@
+        [InlineData(X509KeyStorageFlags.DefaultKeySet, true)]
+        [InlineData(X509KeyStorageFlags.UserKeySet)]
+        [InlineData(X509KeyStorageFlags.UserKeySet, true)]
+        [InlineData(X509KeyStorageFlags.MachineKeySet)]
+        [InlineData(X509KeyStorageFlags.MachineKeySet, true)]
+        public static void AllFilesDeleted_SinglePrivateKey_SingleLoader_KeepCsp(X509KeyStorageFlags storageFlags, bool capi = false)
+        {
+            EnsureNoKeysGained(
+                (Flags: storageFlags, Bytes: MakePfx(1, 5, capi)),
+                state => X509CertificateLoader.LoadPkcs12(state.Bytes, "", state.Flags, s_cspPreservingLimits));
+        }
+
+        [Theory]
+        [InlineData(X509KeyStorageFlags.DefaultKeySet)]
+        [InlineData(X509KeyStorageFlags.DefaultKeySet, true)]
+        [InlineData(X509KeyStorageFlags.UserKeySet)]
+        [InlineData(X509KeyStorageFlags.UserKeySet, true)]
+        [InlineData(X509KeyStorageFlags.MachineKeySet)]
+        [InlineData(X509KeyStorageFlags.MachineKeySet, true)]
+        public static void AllFilesDeleted_MultiplePrivateKey_CollLoader_KeepCsp(X509KeyStorageFlags storageFlags, bool capi = false)
+        {
+            EnsureNoKeysGained(
+                (Flags: storageFlags, Bytes: MakePfx(6, capi: capi)),
+                state => new ImportedCollection(
+                    X509CertificateLoader.LoadPkcs12Collection(state.Bytes, "", state.Flags, s_cspPreservingLimits)));
+        }
+
+        [Theory]
+        [InlineData(X509KeyStorageFlags.DefaultKeySet)]
+        [InlineData(X509KeyStorageFlags.DefaultKeySet, true)]
+        [InlineData(X509KeyStorageFlags.UserKeySet)]
+        [InlineData(X509KeyStorageFlags.UserKeySet, true)]
+        [InlineData(X509KeyStorageFlags.MachineKeySet)]
+        [InlineData(X509KeyStorageFlags.MachineKeySet, true)]
+        public static void AllFilesDeleted_SinglePrivateKey_CollLoader_KeepCsp(X509KeyStorageFlags storageFlags, bool capi = false)
+        {
+            EnsureNoKeysGained(
+                (Flags: storageFlags, Bytes: MakePfx(1, 5, capi)),
+                state => new ImportedCollection(
+                    X509CertificateLoader.LoadPkcs12Collection(state.Bytes, "", state.Flags, s_cspPreservingLimits)));
+        }
+
+        private static byte[] MakePfx(
+            int certAndKeyCount = 1,
+            int certOnlyCount = 0,
+            bool capi = false,
+            [CallerMemberName] string? name = null)
+        {
+            Pkcs12SafeContents keys = new Pkcs12SafeContents();
+            Pkcs12SafeContents certs = new Pkcs12SafeContents();
+            DateTimeOffset notBefore = DateTimeOffset.UtcNow.AddMinutes(-5);
+            DateTimeOffset notAfter = notBefore.AddMinutes(10);
+
+            // Every other key is a machine key. Depending on the minute when this line of code runs, it's
+            // either starting with the first, or with the second.
+            int machineKeySkew = DateTime.UtcNow.Minute % 2;
```

Comment:
Why not `Random.Shared.Next(0, 2)`?

Reply:
Or, a different question - is there a way to make the test behavior deterministic? Non-determinism from time or randomness makes tracking down intermittent failures difficult.

Reply:
The problem is that we select the "first" (which I believe is last here) cert with a private key, so if any particular method always gets that as "machine" or "user" then it's not testing DefaultKeySet both ways.

But I can move the skew to a parameter and just go both ways with DefaultKeySet.